org.eclipse.paho.android.service/src/main/java/org/eclipse/paho/android/service/MqttAndroidClient.java

```diff
@@ -1526,6 +1528,28 @@ private void simpleAction(IMqttToken token, Bundle data) {
 		}
 	}
 
+	private void subscribeResultAction(IMqttToken token, Bundle data) {
+		if (token != null) {
+			int[] grantedQos = data.getIntArray(MqttServiceConstants.GRANTED_QOS);
+			String[] topics = token.getTopics();
+			List<String> successTopics = new ArrayList<>();
+			List<String> failedTopics = new ArrayList<>();
+			if (grantedQos != null) {
+				for (int i = 0; i < topics.length; i++) {
+					if (grantedQos[i] == 128) {
+						failedTopics.add(topics[i]);
+					} else {
+						successTopics.add(topics[i]);
+					}
+				}
+			}
+
+			((MqttTokenAndroid) token).notifySubscription(successTopics, failedTopics);

```

Comment:
should we check `instanceof` here?

Reply:
This function is only called from `subscribeAction` which is inside `onReceive` of SUBSCRIBE_ACTION.. hence have not added any additional checks